File: src/henaxel/raytracer/VolumetricSphere.java
package henaxel.raytracer;

import henaxel.raytracer.materials.Material;
import henaxel.utils.Vec3;

public class VolumetricSphere extends Sphere {
    private double density;

    public VolumetricSphere(Vec3 pos, double radius, Material material, double density) {
        super(pos, radius, material);
        this.density = density;
    }

    @Override
    protected HitResult hit(Ray r, double t_min, double t_max) {
        Vec3 oc = Vec3.subtract( r.origin(), pos);
        double a = Vec3.dot(r.direction(), r.direction());
        double b = Vec3.dot(oc, r.direction());
        double c = Vec3.dot(oc, oc) - radius*radius;
        double discriminant = b*b - a*c;

        //TODO: Fix discrepancies in below code
        if(discriminant > 0) {
                double tInObject = ((Math.random() * 2.0) - 1) * (Math.sqrt(b * b - a * c) / a);
                double t = (-b + tInObject);
                if (t < t_max && t > t_min && Math.random() < density) {
                    Vec3 hitPos = r.pointAtParameter(t);
                    Vec3 n = Vec3.subtract(hitPos, pos).unitVector();
                    Ray scatter = mat.scatter(r, hitPos);
                    return new HitResult(hitPos, n, t, scatter, mat.getAlbedo());
                } else {
                if (t < t_max && t > t_min) {
                    Vec3 hitPos = r.pointAtParameter(t);
                    Vec3 n = Vec3.subtract(hitPos, pos).unitVector();
                    Ray scatter = new Ray(hitPos, r.direction());
                    return new HitResult(hitPos, n, t, scatter, mat.getAlbedo());
                }
            }
        }
        return null;
    }
}

File: src/henaxel/raytracer/VolumetricSphere.kt
package henaxel.raytracer

import henaxel.raytracer.materials.Material
import henaxel.raytracer.utils.Vec3
import kotlin.math.sqrt
import kotlin.random.Random

//TODO: Volumetricsphere probably shouldn't extend Sphere, not much is common between them
class VolumetricSphere(pos: Vec3, radius: Double, material: Material, var density: Double) : Sphere(pos, radius, material) {

    override fun hit(ray: Ray, tMin: Double, tMax: Double): HitResult? {
        val oc = ray.pos - pos
        val a = ray.dir * ray.dir
        val b = oc * ray.dir
        val c = (oc * oc) - super.radius*super.radius
        val discriminant = b*b - a*c

        if (discriminant > 0) {
            val tClosest =  ((-b - sqrt(discriminant)) / a).coerceIn(tMin, tMax)
            val tFarthest = ((-b + sqrt(discriminant)) / a).coerceIn(tMin, tMax)
            if (tClosest == tFarthest) return null

            if ((Random.nextDouble(0.0,1.0)*(tFarthest-tClosest)*density) < 1.0) return null

            val t = Random.nextDouble(tClosest, tFarthest)  //TODO: Implement the ray scatteirng closer to the entry with higher density
            val hitPos = ray.pointAtParameter(t)
            val normal = (hitPos-pos)/super.radius
            val scatter = mat.scatter(ray, hitPos, normal)
            val uv = super.getUV(hitPos)

            return HitResult(hitPos, normal, t, scatter, mat.getAlbedo(uv[0], uv[1]))
        }
        return null
    }
}